tests/returntypes/test_parameter_shift_new.py

```diff
@@ -0,0 +1,2471 @@
+        P = np.array([1])
+        x, y = 0.765, -0.654
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.var(qml.Projector(P, wires=0) @ qml.PauliX(1))
+
+        tape.trainable_params = {0, 1}
+
+        res = dev.execute_new(tape)
+        expected = 0.25 * np.sin(x / 2) ** 2 * (3 + np.cos(2 * y) + 2 * np.cos(x) * np.sin(y) ** 2)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # # circuit jacobians
+        tapes, fn = qml.gradients.param_shift(tape)
+        gradA = fn(dev.batch_execute_new(tapes))
+
+        # TODO: check when finite diff ready:
+        # tapes, fn = qml.gradients.finite_diff(tape)
+        # gradF = fn(dev.batch_execute_new(tapes))
+
+        expected = np.array(
+            [
+                [
+                    0.5 * np.sin(x) * (np.cos(x / 2) ** 2 + np.cos(2 * y) * np.sin(x / 2) ** 2),
+                    -2 * np.cos(y) * np.sin(x / 2) ** 4 * np.sin(y),
+                ]
+            ]
+        )
+        assert np.allclose(gradA, expected, atol=tol, rtol=0)
+        # TODO: check when finite diff ready:
+        # assert gradF == pytest.approx(expected, abs=tol)
+
+    def cost1(x):
+        qml.Rot(*x, wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+    def cost2(x):
+        qml.Rot(*x, wires=0)
+        return [qml.expval(qml.PauliZ(0))]
+
+    def cost3(x):
+        qml.Rot(*x, wires=0)
+        return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
+
+    def cost4(x):
+        qml.Rot(*x, wires=0)
+        return qml.probs([0, 1])
+
+    def cost5(x):
+        qml.Rot(*x, wires=0)
+        return [qml.probs([0, 1])]
+
+    def cost6(x):
+        qml.Rot(*x, wires=0)
+        return [qml.probs([0, 1]), qml.probs([2, 3])]
+
+    costs_and_expected_expval = [
+        (cost1, [3], False),
+        (cost2, [3], True),
+        (cost3, [2, 3], True),
+    ]
+
+    @pytest.mark.parametrize("cost, expected_shape, list_output", costs_and_expected_expval)
+    def test_output_shape_matches_qnode_expval(self, cost, expected_shape, list_output):
+        """Test that the transform output shape matches that of the QNode."""
+        dev = qml.device("default.qubit", wires=4)
+
+        x = np.random.rand(3)
+        circuit = qml.QNode(cost, dev)
+
+        res = qml.gradients.param_shift(circuit)(x)
+        assert isinstance(res, tuple)
+        assert len(res) == expected_shape[0]
+
+        if len(expected_shape) > 1:
+            for r in res:
+                assert isinstance(r, tuple)
+                assert len(r) == expected_shape[1]
+
+    costs_and_expected_probs = [
+        (cost4, [3, 4], False),
+        (cost5, [3, 4], True),
+        # The output shape of transforms for 2D qnode outputs (cost6) is currently
+        # transposed, e.g. (4, 1, 3) instead of (1, 4, 3).
+        # TODO: fix qnode/expected once #2296 is resolved
```

Comment:
Those lines can probably be removed.

Reply:
Not sure about this :thinking: The shapes are off here, but as the QNode decorator requires a revisit, I'd rather have this TODO here for now. Any thoughts?

Reply:
Not sure about this :thinking: The shapes are off here, but as the QNode decorator requires a revisit, I'd rather have this TODO here for now. Any thoughts?

Reply:
Ok you are right, let's use the TODO.